### PR TITLE
Post facto library addition to maven projects

### LINKED ISSUE #1463: Make BuildPath work for maven projects too
I.e. the libraries to add panel

### PR DESCRIPTION
**Pull request by elharo**
fix #1463 

There's already a different maven flow for this, but this should work too.

### CODE REVIEW

**Comment by briandealwis** [comment]
Sporadic failure:
```
com.google.cloud.tools.eclipse.appengine.libraries.AppEngineLibrariesInPluginXmlTest
testObjectifyLibraryConfig(com.google.cloud.tools.eclipse.appengine.libraries.AppEngineLibrariesInPluginXmlTest)  Time elapsed: 0.003 sec  <<< FAILURE!
java.lang.AssertionError: 
Expected: is <http://static.javadoc.io/com.googlecode.objectify/objectify/5.1.14/>
     but: was null
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at com.google.cloud.tools.eclipse.appengine.libraries.AppEngineLibrariesInPluginXmlTest.testObjectifyLibraryConfig(AppEngineLibrariesInPluginXmlTest.java:148)

testAppEngineApiLibraryConfig(com.google.cloud.tools.eclipse.appengine.libraries.AppEngineLibrariesInPluginXmlTest)  Time elapsed: 0 sec  <<< FAILURE!
java.lang.AssertionError: 
Expected: is <https://cloud.google.com/appengine/docs/java/javadoc/>
     but: was null
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at com.google.cloud.tools.eclipse.appengine.libraries.AppEngineLibrariesInPluginXmlTest.testAppEngineApiLibraryConfig(AppEngineLibrariesInPluginXmlTest.java:69)

testEndpointsLibraryConfig(com.google.cloud.tools.eclipse.appengine.libraries.AppEngineLibrariesInPluginXmlTest)  Time elapsed: 0 sec  <<< FAILURE!
java.lang.AssertionError: 
Expected: is <https://cloud.google.com/appengine/docs/java/endpoints/javadoc/>
     but: was null
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at com.google.cloud.tools.eclipse.appengine.libraries.AppEngineLibrariesInPluginXmlTest.testEndpointsLibraryConfig(AppEngineLibrariesInPluginXmlTest.java:113)
```

**Comment by elharo** [comment]
Yes, has anyone else seen that sporadic failure in AppEngineLibrariesInPluginXmlTest on any other PRs? It seems more likely to have been introduced in another recent PR (and I know which one) than in this one. I'd like to confirm.

**Comment by akerekes** [comment]
I haven't seen it before and it seems to fail consistently on Mars.

**Comment by elharo** [comment]
Fails on mars. What's different about that I wonder?

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import org.w3c.dom.ls.LSOutput;
+import org.w3c.dom.ls.LSSerializer;
+import org.xml.sax.SAXException;
+
+import com.google.cloud.tools.eclipse.appengine.libraries.model.Library;
+import com.google.cloud.tools.eclipse.appengine.libraries.model.LibraryFile;
+import com.google.cloud.tools.eclipse.appengine.libraries.model.MavenCoordinates;
+import com.google.common.base.Preconditions;
+
+class Pom {
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java]: Hmm, why not use m2e, like the PomEdits class?

Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java]: Because that's internal, unreliable, undocumented code. It's easier and safer to manipulate the XML directly.

**Comment by chanseokoh** [comment]
> Fails on mars. What's different about that I wonder?

Maybe Mars had a bug that was fixed since Neon, or Neon made an incompatible change on something?

**Comment by elharo** [comment]
I think I know what's happening now. After removing the Javadoc asserts the tests now fail like this:

<<< FAILURE!
java.lang.AssertionError: 
Expected: is "com.googlecode.objectify"
     but: was "com.example.group2"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at com.google.cloud.tools.eclipse.appengine.libraries.AppEngineLibrariesInPluginXmlTest.testObjectifyLibraryConfig(AppEngineLibrariesInPluginXmlTest.java:156)

testServletApiLibraryConfig(com.google.cloud.tools.eclipse.appengine.libraries.AppEngineLibrariesInPluginXmlTest)  Time elapsed: 0 sec
testLibrarySize(com.google.cloud.tools.eclipse.appengine.libraries.AppEngineLibrariesInPluginXmlTest)  Time elapsed: 0 sec
testAppEngineApiLibraryConfig(com.google.cloud.tools.eclipse.appengine.libraries.AppEngineLibrariesInPluginXmlTest)  Time elapsed: 0 sec  <<< FAILURE!
java.lang.AssertionError: 
Expected: is "com.google.appengine"
     but: was "com.example.group0"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at com.google.cloud.tools.eclipse.appengine.libraries.AppEngineLibrariesInPluginXmlTest.testAppEngineApiLibraryConfig(AppEngineLibrariesInPluginXmlTest.java:77)

Those package names (com.group0, group2, etc.) are set up in PomTest so pretty clearly we have some connection between the tests, probably through shared static state in the RegistryFactory or some such.

**Comment by elharo** [comment]
Yes, I see exactly why this is happening. AppEngineLibraries exposes mutable static global state, and PomTest goes right ahead and mutates its. I really should know better than that. Mea culpa. Mea maxima culpa.

**Comment by elharo** [comment]
Travis has seemed exceptionally flaky lately in ways that have nothing to do with our code (e.g. failure to load artifacts on one or more runs) so multiple restarts have been required. However all tests have now passed simultaneously so PTAL @akerekes @briandealwis @chanseokoh 

Note that this PR now locks down a lot of code to make it less public and mutable than it was, since mutating private state is what led to the earlier flaky bugs on Mars that were my fault.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java [hunk 2]
```diff
@@ -89,23 +89,27 @@ public void addDependencies(List<Library> libraries) throws CoreException {
...
-        Element version = document.createElement("version");
-        version.setTextContent(coordinates.getVersion());
-        dependency.appendChild(groupId);
-        dependency.appendChild(artifactId);
-        dependency.appendChild(version);
+        String groupId = coordinates.getGroupId();
+        String artifactId = coordinates.getArtifactId();
         
-        dependencies.appendChild(dependency);
+        if (!dependencyExists(dependencies, groupId, artifactId)) {
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java]: Is there a way to bubble up a warning about the duplicate to the user? The existing dependency may be incompatible with the ones we add (which expected the version that was omitted).
Also at least we should file an issue to consider scope and classifier for the dependency, i.e. a dependency our libraries need may only be present with test scope; or a dependency's javadoc/source is added via a classifier, but not the binary jar. These may be minor corner cases, but we should have at least an issue documenting them as known.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+import com.google.cloud.tools.eclipse.appengine.libraries.model.LibraryFile;
+import com.google.cloud.tools.eclipse.appengine.libraries.model.MavenCoordinates;
+import com.google.cloud.tools.eclipse.test.util.project.TestProjectCreator;
+
+public class PomTest {
+  
+  @Rule public final TestProjectCreator projectCreator = new TestProjectCreator();
+
+  // todo we're doing enough of this we should import or write some utilities
+  private static final DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java]: We could make a JUnit @Rule for it.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  public void testAddDependencies() 
+      throws CoreException, ParserConfigurationException, IOException, SAXException {
+    
+    MavenCoordinates coordinates0 = new MavenCoordinates("com.example.group0", "artifact0");
+    coordinates0.setVersion("1.2.3");
+    LibraryFile file0 = new LibraryFile(coordinates0);
+    List<LibraryFile> list0 = new ArrayList<>();
+    list0.add(file0);
+    
+    List<LibraryFile> list1 = new ArrayList<>();
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java]: It may make sense to extract this and the following 2 lines into a helper method and use it to create all `Library`s/`List<LibraryFile>`s.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    Document actual = parse(contents);
+    
+    NodeList dependencies = actual.getElementsByTagName("dependencies");
+    Assert.assertEquals(1, dependencies.getLength());
+    NodeList children = ((Element) dependencies.item(0)).getElementsByTagName("dependency");
+    
+    Assert.assertEquals(4, children.getLength());
+    
+    Element child0 = (Element) children.item(0);
+    Element groupId = getOnlyChild(child0, "groupId");
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java]: We could extract group ID, artifact ID and version verification into a helper method.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries.ui/src/com/google/cloud/tools/eclipse/appengine/libraries/ui/AppEngineLibrariesPage.java [hunk 6]
```diff
@@ -93,18 +91,20 @@ public void initialize(IJavaProject project, IClasspathEntry[] currentEntries) {
...
       return null;
     }
+
     try {
-      IClasspathEntry[] added =
-          BuildPath.addLibraries(project, libraries, new NullProgressMonitor());
-      return added;
+      if (MavenUtils.hasMavenNature(project.getProject())) {
+        BuildPath.addMavenLibraries(project.getProject(), libraries, new NullProgressMonitor());
+        return new IClasspathEntry[0]; // todo return something
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries.ui/src/com/google/cloud/tools/eclipse/appengine/libraries/ui/AppEngineLibrariesPage.java]: Do we need an issue for this TODO or is returning an empty array fine?

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java [hunk 7]
```diff
@@ -51,8 +57,33 @@ public static void addLibraries(IProject project, List<Library> libraries,
...
+      addLibraries(javaProject, libraries, monitor);
+    }
+  }
+
+  public static void addMavenLibraries(IProject project, List<Library> libraries,
+      IProgressMonitor monitor) throws CoreException {
+    // see m2e-core/org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/actions/AddDependencyAction.java
+    // m2e-core/org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/editing/AddDependencyOperation.java
+    
+    // todo use monitor
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java]: Is there a corresponding issue?

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java [hunk 8]
```diff
@@ -51,8 +57,33 @@ public static void addLibraries(IProject project, List<Library> libraries,
...
+    SubMonitor subMonitor = SubMonitor.convert(monitor,
+        Messages.getString("adding.app.engine.libraries"), libraries.size()); //$NON-NLS-1$
+    
+    IFile pomFile = project.getFile("pom.xml");
+    
+    try {
+      Pom pom = Pom.parse(pomFile);
+      pom.addDependencies(libraries);
+    } catch (SAXException | IOException ex) {
+      IStatus status = new Status(IStatus.ERROR,
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java]: You could use StatusUtil.error() to be consistent.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+  static Pom parse(IFile pomFile) throws SAXException, IOException, CoreException {
+    Preconditions.checkState(pomFile.exists(), pomFile.getFullPath() + " does not exist");
+    
+    try {
+      DocumentBuilder builder = builderFactory.newDocumentBuilder();
+      Document document = builder.parse(pomFile.getContents());
+      Pom pom = new Pom(document, pomFile);
+      return pom;
+    } catch (ParserConfigurationException ex) {
+      IStatus status = new Status(IStatus.ERROR,
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java]: You could use StatusUtil.error() to be consistent.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+          "com.google.cloud.tools.eclipse.appengine.libraries", ex.getMessage(), ex);
+      throw new CoreException(status);
+    }
+  }
+
+  void addDependencies(List<Library> libraries) throws CoreException {
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java]: We should add comments to indicate that these libraries have been added by our plugin (or rather file an issue and resolve it in a separate PR)

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java (hunk 1 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java]: If we used m2e and PomEdits, would that help in identifying if the parent pom contains the given dependency not the one that we're editing? I.e. checking the effective pom instead of the actual file.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/LibraryFile.java [hunk 11]
```diff
@@ -39,6 +39,7 @@ public LibraryFile(MavenCoordinates mavenCoordinates) {
     this.mavenCoordinates = mavenCoordinates;
   }
 
+  // todo Danger! This exposes mutable internal state
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/LibraryFile.java]: Can you file an issue for this?

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java (hunk 3 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java]: That would only work in JUnit tests. We need this in the model code too.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java (hunk 4 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java]: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1485

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java (hunk 5 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java]: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1485

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries.ui/src/com/google/cloud/tools/eclipse/appengine/libraries/ui/AppEngineLibrariesPage.java (hunk 6 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries.ui/src/com/google/cloud/tools/eclipse/appengine/libraries/ui/AppEngineLibrariesPage.java]: It is. Done.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java (hunk 7 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java]: This is quick enough a monitor doesn't seem to help; removed.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java (hunk 8 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java]: done

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java (hunk 1 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java]: Not sure; but either way it's still internal code we can't rely on.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java (hunk 9 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java]: done

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java (hunk 10 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java]: I'm not sure we should clutter pom.xml with comments about tooling. We can hash this out offline if you like.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/LibraryFile.java (hunk 11 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/LibraryFile.java]: done

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java (hunk 2 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java]: A duplicate is not necessarily a mistake. E.g. this will catch cases where multiple libraries all add the same guava.jar.

Versions worth considering though I'm not sure what to do. Just pick the latest even though existing code may require the older version? We may need to leave this to the user to sort out. 

For scope, filed https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1486

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java [hunk 12]
```diff
@@ -70,18 +70,14 @@ public static void addMavenLibraries(IProject project, List<Library> libraries,
...
-    
     IFile pomFile = project.getFile("pom.xml");
     
     try {
       Pom pom = Pom.parse(pomFile);
       pom.addDependencies(libraries);
     } catch (SAXException | IOException ex) {
-      IStatus status = new Status(IStatus.ERROR,
-          "com.google.cloud.tools.eclipse.appengine.libraries", ex.getMessage(), ex);
+      IStatus status = StatusUtil.error("com.google.cloud.tools.eclipse.appengine.libraries",
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java]: This probably isn't going to work as expected. The first parameter `origin` is either the instance of the class that is the origin for the error or the class itself. So it probably should be either `this` or `this.getClass()` (or `BuildPath.class`, but moving this code another class during a refactor would require changing it)

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java [hunk 13]
```diff
@@ -74,8 +74,8 @@ static Pom parse(IFile pomFile) throws SAXException, IOException, CoreException
       Pom pom = new Pom(document, pomFile);
       return pom;
     } catch (ParserConfigurationException ex) {
-      IStatus status = new Status(IStatus.ERROR,
-          "com.google.cloud.tools.eclipse.appengine.libraries", ex.getMessage(), ex);
+      IStatus status = StatusUtil.error("com.google.cloud.tools.eclipse.appengine.libraries",
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java]: Probably won't work as expected, see above.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java (hunk 2 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java]: Yes, I also thought to notify the user if the versions are different. I also would leave it to the user to figure out, but making her aware of the situation would be helpful.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java (hunk 12 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java]: done

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java (hunk 13 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java]: done
